### PR TITLE
Add custom 'logging.Logger' that has warn_once(), debug_once(), etc m…

### PR DESCRIPTION
**Pull request by bellecarrell**
Addresses #3606 by creating custom Logger and Filter classes.

### CODE REVIEW

**Comment by bellecarrell** [comment]
Yes, that is simpler. I made the changes.

File: allennlp/common/logging.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    """
+
+    def __init__(self, name):
+        super().__init__(name)
+        self.msgs = set()
+
+    def debug_once(self, msg, *args, **kwargs):
+        if msg not in self.msgs:
+            self.debug(msg, *args, **kwargs)
+        self.msgs.add(msg)
```
Comment (dirkgr) [inline allennlp/common/logging.py:21]: You can indent the `set.add` call as well, for a minuscule performance boost.

File: allennlp/common/logging.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+class AllenNlpLogger(logging.Logger):
+    """
+    A custom subclass of 'logging.Logger' that keeps a set of messages to
+    implement {debug,info,etc.}_once() methods.
+    """
+
+    def __init__(self, name):
+        super().__init__(name)
+        self.msgs = set()
```
Comment (dirkgr) [inline allennlp/common/logging.py:16]: Can you call this variable `self._seen_msgs`, or something with an underscore? We generally use underscores to mark "private" variables.

File: allennlp/tests/common/logging_test.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+    def setUp(self):
+        super().setUp()
+        logger = logging.getLogger(str(self.__class__.i))
+        self.test_log_file = os.path.join(self.TEST_DIR, "test.log")
+        hdlr = logging.FileHandler(self.test_log_file)
+        logger.addHandler(hdlr)
+        self.logger = logger
+        self.msg = "test message"
+        self.__class__.i += 1
```
Comment (dirkgr) [inline allennlp/tests/common/logging_test.py:18]: What is this for?

File: allennlp/tests/common/logging_test.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
+import os
+import logging
+from allennlp.common.logging import AllenNlpLogger
+from allennlp.common.testing import AllenNlpTestCase
+
+
+class TestLogging(AllenNlpTestCase):
```
Comment (dirkgr) [inline allennlp/tests/common/logging_test.py:7]: Can you add a test that makes sure that this sequence only produces one log entry:
```Python
logger.debug_once("There are %d lights.", 4)
logger.debug_once("There are %d lights.", 5)
```
The reason I am concerned about it is that if this doesn't work, we might end up storing millions of messages in `self.msg`.

Comment (dirkgr) [reply allennlp/tests/common/logging_test.py:7]: In fact, maybe add that to the test, i.e., make sure that after that sequence, `sizeof(logger.msg)` is 1.

**Pull request merged**